File: 2021/day2Dive/day2.mjs
// --- Day 2: Dive! ---

import { readFileSync } from "node:fs";

const input = "day2Input.txt";

const plannedCourse = readFileSync(input, { encoding: "utf-8" })
  .split("\n")
  .map((step) => step.split(" "));

let position = { horizontalPosition: 0, depth: 0 };

let stepsMechanism = {
  forward: { direction: "horizontalPosition", sense: 1 },
  up: { direction: "depth", sense: -1 },
  down: { direction: "depth", sense: 1 },
};

plannedCourse.forEach(([stepDirection, stepValue]) => {
  position[stepsMechanism[stepDirection].direction] +=
    parseInt(stepValue) * stepsMechanism[stepDirection].sense;
});

console.log(
  `After following the planned course:\n Horizontal position: ${
    position.horizontalPosition
  }\n Depth: ${position.depth}\n Product: ${
    position.horizontalPosition * position.depth
  }`
); // Answer Part 1

//Part 2

position = { horizontalPosition: 0, depth: 0, aim: 0 };

stepsMechanism = {
  forward: { direction: "horizontalPosition", sense: 1, depthFactor: 1 },
  up: { direction: "aim", sense: -1, depthFactor: 0 },
  down: { direction: "aim", sense: 1, depthFactor: 0 },
};

plannedCourse.forEach(([stepDirection, stepValue]) => {
  position[stepsMechanism[stepDirection].direction] +=
    parseInt(stepValue) * stepsMechanism[stepDirection].sense;

  position.depth +=
    position.aim *
    parseInt(stepValue) *
    stepsMechanism[stepDirection].depthFactor;
});

console.log(
  `After following the the new instructions:\n Horizontal position: ${
    position.horizontalPosition
  }\n Depth: ${position.depth}\n Product: ${
    position.horizontalPosition * position.depth
  }`
); // Answer Part 2

File: 2021/day2Dive/day2.py
# --- Day 2: Dive! ---

input = "day2Input.txt"

with open(input, "r", encoding="utf-8") as file:
    planned_course = [step.split() for step in file.read().split("\n")]

position = {"horizontal_position": 0, "depth": 0}


steps_mechanism = {
    "forward": {"direction": "horizontal_position", "sense": 1},
    "up": {"direction": "depth", "sense": -1},
    "down": {"direction": "depth", "sense": 1},
}

for step_direction, step_value in planned_course:
    position[steps_mechanism[step_direction]["direction"]] += (
        int(step_value) * steps_mechanism[step_direction]["sense"]
    )

print(
    f"After following the planned course:\n Horizontal position: {position['horizontal_position']}\n Depth: {position['depth']}\n Product: {position['horizontal_position']*position['depth']}"
)  # Answer Part 1

# Part 2

position = {"horizontal_position": 0, "depth": 0, "aim": 0}

steps_mechanism = {
    "forward": {"direction": "horizontal_position", "sense": 1, "depth_factor": 1},
    "up": {"direction": "aim", "sense": -1, "depth_factor": 0},
    "down": {"direction": "aim", "sense": 1, "depth_factor": 0},
}

for step_direction, step_value in planned_course:
    position[steps_mechanism[step_direction]["direction"]] += (
        int(step_value) * steps_mechanism[step_direction]["sense"]
    )
    position["depth"] += (
        position["aim"] * int(step_value) * steps_mechanism[step_direction]["depth_factor"]
    )

print(
    f"After following the new instructions:\n Horizontal position: {position['horizontal_position']}\n Depth: {position['depth']}\n Product: {position['horizontal_position']*position['depth']}"
)  # Answer Part 2
